File: query.py
'''Very simple querying of the output matrix to find nearest neighbours
'''

import argparse

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--guid", required=True, help="GUID of the sample to query for")
    parser.add_argument("--snps", required=False, default=float('inf'), help="SNP threshold for nearest neighbour")
    options = parser.parse_args()

    guid = options.guid
    cutoff = float(options.snps)

    with open("outputs/all.txt") as f:
        data = [line.strip() for line in f]
    
    matches = {}
    for line in data:
        guid1, guid2, snp = line.split(" ")
        if guid1 == guid:
            snp = int(snp)
            if snp <= cutoff:
                matches[guid2] = snp
        elif guid2 == guid:
            snp = int(snp)
            if snp <=cutoff:
                matches[guid1] = snp
    
    print(matches)

File: query.cpp
#include <string>
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;

int main(int nargs, const char* args[]){
    //Must use as `./query <guid> <cutoff>`
    if(nargs != 3){
        cout << "Incorrect usage. " << endl;
        return 1;
    }

    string guid = args[1];
    int cutoff = stoi(args[2]);
    
    char ch;
    vector<string> line;
    string acc;
    fstream fin("outputs/all.txt", fstream::in);
    int i = 0;
    while (fin >> noskipws >> ch) {
        i++;
        if(ch == ' '){
            line.push_back(acc);
            acc = "";
            continue;
        }
        else if(ch == '\n'){
            //We've parsed a line, so convert and compare
            if(line.size() != 2){
                cout << "??? Line " << i << ": " << line.size() << endl;
                continue;
            }
            string guid1 = line.at(0);
            string guid2 = line.at(1);
            int snp = stoi(acc);
            if(guid == guid1 && snp <= cutoff){
                cout << guid1 << ": " << snp << endl;
            }
            if(guid == guid2 && snp <= cutoff){
                cout << guid2 << ": " << snp << endl;
            }

            line = {};
            acc = "";

        }
        else{
            acc += ch;
        }
    }
    fin.close();
}